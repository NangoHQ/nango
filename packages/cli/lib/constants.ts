export const TYPES_FILE_NAME = 'models.ts';
export const exampleSyncName = 'github-issue-example';

export const port = process.env['NANGO_PORT'] || process.env['SERVER_PORT'] || '3003';
export const localhostUrl = `http://localhost:${port}`;
export const cloudHost = 'https://api.nango.dev';
