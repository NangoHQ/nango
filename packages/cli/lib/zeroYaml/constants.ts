import ts from 'typescript';

export const npmPackageRegex = /^[^./\s]/;
export const importRegex = /^import ['"](?<path>\.\/[^'"]+)['"];/gm;

export const tsconfig: ts.CompilerOptions = {
    module: ts.ModuleKind.CommonJS,
    target: ts.ScriptTarget.ESNext,
    strict: true,
    esModuleInterop: true,
    skipLibCheck: true,
    forceConsistentCasingInFileNames: true,
    moduleResolution: ts.ModuleResolutionKind.NodeJs,
    allowUnusedLabels: false,
    allowUnreachableCode: false,
    exactOptionalPropertyTypes: true,
    noFallthroughCasesInSwitch: true,
    noImplicitOverride: true,
    noImplicitReturns: true,
    noPropertyAccessFromIndexSignature: true,
    noUncheckedIndexedAccess: true,
    noUnusedLocals: true,
    noUnusedParameters: true,
    declaration: false,
    sourceMap: true,
    composite: false,
    checkJs: false,
    noEmit: true
};
export const tsconfigString: Record<string, any> = {
    ...tsconfig,
    module: 'commonjs',
    target: 'esnext',
    importsNotUsedAsValues: 'remove',
    jsx: 'react',
    moduleResolution: 'node16'
};

export type BabelErrorType = 'nango_unsupported_export' | 'nango_invalid_function_param' | 'nango_invalid_default_export' | 'nango_invalid_export_constant';
export class BabelError extends Error {
    type;
    constructor(type: BabelErrorType) {
        super(type);
        this.type = type;
    }
}

const defaultMsg = 'Invalid default export: should be createAction(), createSync() or createOnEvent()';
export const customErrors: Record<BabelErrorType, string> = {
    nango_unsupported_export: defaultMsg,
    nango_invalid_function_param: 'Invalid function parameter, should be an object',
    nango_invalid_default_export: defaultMsg,
    nango_invalid_export_constant: defaultMsg
};
export class CompileError extends Error {
    type: BabelErrorType | 'failed_to_build_unknown';
    msg: string;
    filePath?: string;
    constructor(type: BabelErrorType | 'failed_to_build_unknown', msg: string) {
        super(msg);
        this.type = type;
        this.msg = msg;
    }
}
