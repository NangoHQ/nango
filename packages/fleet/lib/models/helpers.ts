import crypto from 'crypto';

export function generateImage(): string {
    const charset = '0123456789abcdef';
    const length = 40;
    const randomBytes = new Uint8Array(length);
    crypto.getRandomValues(randomBytes);

    const commitHash = Array.from(randomBytes)
        .map((byte) => charset[byte % charset.length])
        .join('');
    return `generated/image:${commitHash}`;
}
