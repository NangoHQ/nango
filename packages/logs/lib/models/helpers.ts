import { nanoid } from '@nangohq/utils';
import type { ConcatOperationList, MessageRow, MessageRowInsert, OperationRow, OperationRowInsert } from '@nangohq/types';
import { z } from 'zod';
import type { estypes } from '@elastic/elasticsearch';
import { defaultOperationExpiration } from '../env.js';
import type { LogContext } from '../client.js';

export const operationIdRegex = z.string().regex(/([0-9]|[a-zA-Z0-9]{20})/);

export interface FormatMessageData {
    account?: { id: number; name: string };
    user?: { id: number } | undefined;
    environment?: { id: number; name: string } | undefined;
    connection?: { id: number; name: string } | undefined;
    integration?: { id: number; name: string; provider: string } | undefined;
    syncConfig?: { id: number; name: string } | undefined; // TODO: rename to script or something similar because it also apply to actions and on-events scripts
    meta?: MessageRow['meta'];
}

export function getFormattedOperation(
    data: OperationRowInsert,
    { account, user, environment, integration, connection, syncConfig, meta }: FormatMessageData = {}
): OperationRow {
    return {
        ...getFormattedMessage(data as unknown as MessageRow),
        message: operationTypeToMessage[`${data.operation.type}:${data.operation.action}` as ConcatOperationList],
        id: data.id || nanoid(),
        operation: data.operation,

        accountId: account?.id ?? data.accountId ?? -1,
        accountName: account?.name || data.accountName || '',

        environmentId: environment?.id ?? data.environmentId ?? null,
        environmentName: environment?.name || data.environmentName || null,

        integrationId: integration?.id ?? data.integrationId ?? null,
        integrationName: integration?.name || data.integrationName || null,
        providerName: integration?.provider || data.providerName || null,

        connectionId: connection?.id ?? data.connectionId ?? null,
        connectionName: connection?.name || data.connectionName || null,

        syncConfigId: syncConfig?.id || data.syncConfigId || null,
        syncConfigName: syncConfig?.name || data.syncConfigName || null,

        jobId: data.jobId || null,
        meta: meta || data.meta || null,

        userId: user?.id || data.userId || null,
        parentId: null
    };
}
export function getFormattedMessage(data: Partial<MessageRow>, { meta }: FormatMessageData = {}): MessageRow {
    const now = new Date();
    return {
        id: data.id || nanoid(), // This ID is for debugging purpose, not for insertion

        source: data.source || 'internal',
        level: data.level || 'info',
        operation: data.operation || null,
        type: data.type || 'log',
        message: data.message || '',
        title: data.title || null,
        code: data.code || null,
        state: data.state || 'waiting',

        accountId: null,
        accountName: null,

        environmentId: null,
        environmentName: null,

        integrationId: null,
        integrationName: null,
        providerName: null,

        connectionId: null,
        connectionName: null,

        syncConfigId: null,
        syncConfigName: null,

        jobId: data.jobId || null,

        userId: null,
        parentId: data.parentId || null,

        error: data.error || null,
        request: data.request || null,
        response: data.response || null,
        meta: meta || data.meta || null,

        createdAt: data.createdAt || now.toISOString(),
        updatedAt: data.updatedAt || now.toISOString(),
        startedAt: data.startedAt || null,
        endedAt: data.endedAt || null,
        expiresAt: data.operation ? data.expiresAt || defaultOperationExpiration.sync() : null
    };
}

// TODO: remove once not used by persist anymore
export const oldLevelToNewLevel = {
    debug: 'debug',
    info: 'info',
    warn: 'warn',
    error: 'error',
    verbose: 'debug',
    silly: 'debug',
    http: 'info'
} as const;

export function getFullIndexName(prefix: string, createdAt: string) {
    return `${prefix}.${new Date(createdAt).toISOString().split('T')[0]}`;
}

export function createCursor({ sort }: estypes.SearchHit): string {
    return Buffer.from(JSON.stringify(sort)).toString('base64');
}

export function parseCursor(str: string): any[] {
    return JSON.parse(Buffer.from(str, 'base64').toString('utf8'));
}

export const operationTypeToMessage: Record<ConcatOperationList, string> = {
    'action:run': 'Action',
    'admin:impersonation': 'Admin logged into another account',
    'auth:create_connection': 'Connection created',
    'auth:post_connection': 'post connection execution',
    'auth:refresh_token': 'Token refreshed',
    'auth:connection_test': 'Connection test',
    'deploy:custom': 'Deploys',
    'deploy:prebuilt': 'Deploys',
    'proxy:call': 'Proxy',
    'sync:cancel': 'Sync execution canceled',
    'sync:init': 'Sync initialized',
    'sync:pause': 'Sync schedule paused',
    'sync:request_run_full': 'Full execution triggered',
    'sync:request_run': 'Incremental execution triggered',
    'sync:run': 'Sync executed',
    'sync:unpause': 'Sync schedule resumed',
    'webhook:incoming': 'External webhook executed',
    'webhook:forward': 'External webhook forwarded',
    'webhook:sync': 'Sync completion webhooks',
    'webhook:connection_create': 'Connection creation webhooks',
    'webhook:connection_refresh': 'Token refresh webhooks',
    'events:post_connection_creation': 'Event-based executions',
    'events:pre_connection_deletion': 'Event-based executions'
};

/**
 * Send buffered logs to elasticsearch
 * Ultimately it would be better to have LogContextBuffer (or an option in LogContext)
 */
export async function flushLogsBuffer(logs: MessageRowInsert[], logCtx: LogContext) {
    await Promise.all(
        logs.map(async (log) => {
            await logCtx.log(log);
        })
    );
}
