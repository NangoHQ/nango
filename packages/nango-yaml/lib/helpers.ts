import path from 'node:path';
import ms from 'ms';
import type { StringValue } from 'ms';
import type { HTTP_METHOD, NangoSyncEndpointV2, NangoYaml, NangoYamlParsed, NangoYamlParsedIntegration, NangoYamlV2Endpoint } from '@nangohq/types';

interface IntervalResponse {
    interval: StringValue;
    offset: number;
}

export function determineVersion(configData: NangoYaml): 'v1' | 'v2' {
    if (!configData.integrations) {
        return 'v1';
    }

    const keys = Object.keys(configData.integrations);
    if (keys.length <= 0) {
        // Actually not possible to know
        return 'v1';
    }

    const firstProviderConfig = configData.integrations[keys[0]!];

    if (firstProviderConfig && ('syncs' in firstProviderConfig || 'actions' in firstProviderConfig || 'on-events' in firstProviderConfig)) {
        return 'v2';
    } else {
        return 'v1';
    }
}

export function getOffset(interval: StringValue, date: Date): number {
    const intervalMilliseconds = ms(interval);

    const nowMilliseconds = date.getMinutes() * 60 * 1000 + date.getSeconds() * 1000 + date.getMilliseconds();

    const offset = nowMilliseconds % intervalMilliseconds;

    if (isNaN(offset)) {
        return 0;
    }

    return offset;
}

/**
 * Get Interval
 * @desc get the interval based on the runs property in the yaml. The offset
 * should be the amount of time that the interval should be offset by.
 * If the time is 1536 and the interval is 30m then the next time the sync should run is 1606
 * and then 1636 etc. The offset should be based on the interval and should never be
 * greater than the interval
 */
export function getInterval(runs: string, date: Date): IntervalResponse | Error {
    if (runs === 'every half day') {
        const response: IntervalResponse = { interval: '12h', offset: getOffset('12h', date) };
        return response;
    }

    if (runs === 'every half hour') {
        const response: IntervalResponse = { interval: '30m', offset: getOffset('30m', date) };
        return response;
    }

    if (runs === 'every quarter hour') {
        const response: IntervalResponse = { interval: '15m', offset: getOffset('15m', date) };
        return response;
    }

    if (runs === 'every hour') {
        const response: IntervalResponse = { interval: '1h', offset: getOffset('1h', date) };
        return response;
    }

    if (runs === 'every day') {
        const response: IntervalResponse = { interval: '1d', offset: getOffset('1d', date) };
        return response;
    }

    if (runs === 'every month') {
        const response: IntervalResponse = { interval: '30d', offset: getOffset('30d', date) };
        return response;
    }

    if (runs === 'every week') {
        const response: IntervalResponse = { interval: '1w', offset: getOffset('1w', date) };
        return response;
    }

    const interval = runs.replace('every ', '') as StringValue;

    if (!ms(interval)) {
        return new Error('sync_interval_invalid');
    }

    if (ms(interval) < ms('30s')) {
        return new Error('sync_interval_too_short');
    }

    const offset = getOffset(interval, date);
    const response: IntervalResponse = { interval, offset };

    return response;
}

export const JAVASCRIPT_AND_TYPESCRIPT_TYPES = {
    primitives: ['string', 'number', 'boolean', 'bigint', 'symbol'],
    builtInObjects: ['Object', 'Array', 'Function', 'Date', 'RegExp', 'Map', 'Set', 'WeakMap', 'WeakSet', 'Promise', 'Symbol', 'Error'],
    utilityTypes: ['Record', 'Partial', 'Readonly', 'Pick', 'Omit', 'Awaited', 'Required', 'Exclude', 'Extract', 'Uppercase', 'Lowercase']
};
const typesLowercase = Object.values(JAVASCRIPT_AND_TYPESCRIPT_TYPES)
    .flat()
    .map((v) => v.toLocaleLowerCase());
const typesWithGenerics = [...JAVASCRIPT_AND_TYPESCRIPT_TYPES.builtInObjects, ...JAVASCRIPT_AND_TYPESCRIPT_TYPES.utilityTypes];
// Only used externally
export function isJsOrTsType(type?: string): boolean {
    if (!type) {
        return false;
    }

    const baseType = type.replace(/\[\]$/, '').toLocaleLowerCase();
    if (typesLowercase.includes(baseType)) {
        return true;
    }

    const genericTypeRegex = new RegExp(`^(${typesWithGenerics.join('|')})<.+>$`, 'i');

    return genericTypeRegex.test(baseType);
}

export const typesAliases: Record<string, string> = {
    integer: 'number',
    int: 'number',
    char: 'string',
    varchar: 'string',
    float: 'number',
    bool: 'boolean',
    string: 'string',
    number: 'number',
    boolean: 'boolean',
    bigint: 'bigint',
    date: 'Date',
    object: 'Record<string, any>',
    any: 'any',
    array: 'any[]',
    undefined: 'undefined'
};

export function getPotentialTypeAlias(value: string): string | undefined {
    return typesAliases[value.toLocaleLowerCase()];
}

export function getNativeDataType(value: string): number | boolean | null | 'undefined' | Error {
    const int = parseInt(value, 10);
    if (!Number.isNaN(int)) {
        return int;
    }

    switch (value.toLocaleLowerCase()) {
        case 'true':
            return true;
        case 'false':
            return false;
        case 'null':
            return null;
        case 'undefined':
            return 'undefined';
        default:
            return new Error();
    }
}

export function isDisallowedType(value: string) {
    return typesWithGenerics.some((type) => value.startsWith(`${type}<`));
}

const regQuote = /^[a-zA-Z0-9_]+$/;
export function shouldQuote(name: string) {
    return !regQuote.test(name);
}

export function getProviderConfigurationFromPath({ filePath, parsed }: { filePath: string; parsed: NangoYamlParsed }): NangoYamlParsedIntegration | null {
    const pathSegments = filePath.split(path.sep);
    const scriptType = pathSegments.length > 1 ? pathSegments[pathSegments.length - 2] : null;
    const isNested = scriptType && ['syncs', 'actions', 'post-connection-scripts', 'on-events'].includes(scriptType);

    const baseName = path.basename(filePath, '.ts');
    let providerConfiguration: NangoYamlParsedIntegration | null = null;
    if (isNested) {
        const providerConfigKey = pathSegments[pathSegments.length - 3];
        providerConfiguration = parsed.integrations.find((config) => config.providerConfigKey === providerConfigKey) || null;
    } else {
        providerConfiguration = parsed.integrations.find((config) => [...config.syncs, ...config.actions].find((sync) => sync.name === baseName)) || null;
    }

    return providerConfiguration;
}

export function parseEndpoint(rawEndpoint: string | NangoSyncEndpointV2 | NangoYamlV2Endpoint, defaultMethod: HTTP_METHOD): NangoSyncEndpointV2 {
    if (typeof rawEndpoint === 'string') {
        const endpoint = rawEndpoint.split(' ');
        if (endpoint.length > 1) {
            return { method: endpoint[0] as HTTP_METHOD, path: endpoint[1] as string };
        }

        return { method: defaultMethod, path: endpoint[0] as string };
    }

    return { method: rawEndpoint.method || defaultMethod, path: rawEndpoint.path, group: rawEndpoint.group };
}
