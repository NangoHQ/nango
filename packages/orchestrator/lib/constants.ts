export const serverRequestSizeLimit = '10mb';
