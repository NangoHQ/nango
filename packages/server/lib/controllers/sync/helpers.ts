export function normalizeSyncParams(syncs: (string | { name: string; variant: string })[]): { syncName: string; syncVariant: string }[] {
    return syncs.map((sync) => {
        if (typeof sync === 'string') {
            if (sync.includes('::')) {
                const [name, variant] = sync.split('::');
                return { syncName: name ?? '', syncVariant: variant ?? '' };
            }
            return { syncName: sync, syncVariant: 'base' };
        }

        return { syncName: sync.name, syncVariant: sync.variant };
    });
}
