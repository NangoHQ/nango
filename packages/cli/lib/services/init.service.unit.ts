import { expect, describe, it, afterEach, vi } from 'vitest';
import path, { join } from 'node:path';
import fs from 'node:fs';
import { removeVersion, getTestDirectory } from '../tests/helpers.js';
import { init } from './init.service.js';
import { exampleSyncName } from '../constants.js';

describe('init', () => {
    const consoleMock = vi.spyOn(console, 'log').mockImplementation(() => undefined);

    afterEach(() => {
        consoleMock.mockReset();
    });

    it('should init the expected files in the nango-integrations directory', async () => {
        const dir = await getTestDirectory('init');
        init({ absolutePath: path.resolve(dir), debug: false });
        expect(fs.existsSync(join(dir, '.nango'))).toBe(true);
        expect(fs.existsSync(join(dir, '.nango/.gitkeep'))).toBe(true);
        expect(fs.existsSync(join(dir, `demo-github-integration/syncs/${exampleSyncName}.ts`))).toBe(true);
        expect(fs.existsSync(join(dir, '.env'))).toBe(true);
        expect(fs.existsSync(join(dir, 'nango.yaml'))).toBe(true);
        expect(fs.existsSync(join(dir, 'models.ts'))).toBe(true);
        expect(removeVersion(fs.readFileSync(join(dir, '.nango/schema.ts')).toString())).toMatchSnapshot();
        expect(removeVersion(fs.readFileSync(join(dir, '.nango/schema.json')).toString())).toMatchSnapshot();
    });

    it('should not overwrite existing integration files', async () => {
        const dir = await getTestDirectory('overwrite');
        init({ absolutePath: dir, debug: false });
        await fs.promises.writeFile(join(dir, `${exampleSyncName}.ts`), 'dummy fake content', 'utf8');

        const dummyContent = 'This is dummy content. Do not overwrite!';
        const exampleFilePath = join(dir, `${exampleSyncName}.ts`);
        await fs.promises.writeFile(exampleFilePath, dummyContent, 'utf8');

        init({ absolutePath: dir });

        expect(fs.existsSync(exampleFilePath)).toBe(true);
        const fileContentAfterInit = await fs.promises.readFile(exampleFilePath, 'utf8');
        expect(fileContentAfterInit).toBe(dummyContent);
    });
});
