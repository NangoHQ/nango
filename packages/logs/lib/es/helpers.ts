import { isTest } from '@nangohq/utils';
import { envs } from '../env.js';
import { logger } from '../utils.js';
import { client } from './client.js';
import { getDailyIndexPipeline, indexMessages, policyRetention } from './schema.js';
import { createMessage } from '../models/messages.js';
import { getFormattedMessage } from '../models/helpers.js';

export async function start() {
    if (!envs.NANGO_LOGS_ENABLED) {
        logger.warning('Elasticsearch is disabled, skipping');
        return;
    }

    logger.info('ðŸ”„ Elasticsearch service starting...');

    await migrateMapping();

    logger.info('âœ… Elasticsearch');
}

export async function migrateMapping() {
    try {
        const index = indexMessages;

        logger.info(`Migrating index "${index.index}"...`);

        // -- Policy
        logger.info(`  Updating policy`);
        await client.ilm.putLifecycle(policyRetention());

        // -- Index
        const existsTemplate = await client.indices.existsIndexTemplate({ name: `${index.index}-template` });
        logger.info(`  ${existsTemplate ? 'updating' : 'creating'} index template "${index.index}"...`);

        await client.indices.putIndexTemplate({
            name: `${index.index}-template`,
            index_patterns: `${index.index}.*`,
            template: {
                settings: index.settings!,
                mappings: index.mappings!,
                aliases: { [index.index]: {} }
            }
        });

        // -- Pipeline
        // Pipeline will automatically create an index based on a field
        // In our case we create a daily index based on "createdAt"
        logger.info(`  Updating pipeline`);
        await client.ingest.putPipeline(getDailyIndexPipeline(index.index));

        const existsAlias = await client.indices.exists({ index: index.index });
        if (!existsAlias) {
            // insert a dummy record to create first index
            logger.info(`  Inserting dummy record`);
            await createMessage(getFormattedMessage({ parentId: '-1' }));
        }
    } catch (err) {
        logger.error(err);
        throw new Error('failed_to_init_elasticsearch');
    }
}

export async function deleteIndex({ prefix }: { prefix: string }) {
    if (!isTest) {
        throw new Error('Trying to delete stuff in prod');
    }

    try {
        const indices = await client.cat.indices({ format: 'json' });
        await Promise.all(
            indices.map(async (index) => {
                if (!index.index?.startsWith(prefix)) {
                    return;
                }

                await client.indices.delete({ index: index.index, ignore_unavailable: true });
            })
        );
    } catch (err) {
        logger.error(err);
        throw new Error('failed_to_deleteIndex');
    }
}
