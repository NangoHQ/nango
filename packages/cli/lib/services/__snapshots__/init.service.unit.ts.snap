// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`init > should init the expected files in the nango-integrations directory 1`] = `
"// ---------------------------
// This file was generated by Nango (vTest)
// You can version this file
// ---------------------------

export interface GithubIssue {
  id: number;
  owner: string;
  repo: string;
  issue_number: number;
  title: string;
  author: string;
  author_id: string;
  state: string;
  date_created: Date;
  date_last_modified: Date;
};
"
`;

exports[`init > should init the expected files in the nango-integrations directory 2`] = `
"{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "GithubIssue": {
      "type": "object",
      "properties": {
        "id": {
          "type": "number"
        },
        "owner": {
          "type": "string"
        },
        "repo": {
          "type": "string"
        },
        "issue_number": {
          "type": "number"
        },
        "title": {
          "type": "string"
        },
        "author": {
          "type": "string"
        },
        "author_id": {
          "type": "string"
        },
        "state": {
          "type": "string"
        },
        "date_created": {
          "type": "string",
          "format": "date-time"
        },
        "date_last_modified": {
          "type": "string",
          "format": "date-time"
        }
      },
      "required": [
        "id",
        "owner",
        "repo",
        "issue_number",
        "title",
        "author",
        "author_id",
        "state",
        "date_created",
        "date_last_modified"
      ],
      "additionalProperties": false
    }
  },
  "$comment": "This file was generated by Nango (vTest)"
}"
`;
